kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreationServiceTest.java

```diff
@@ -0,0 +1,492 @@
+
+    ReplicaCreationService replicaCreationService =
+        new ReplicaCreationService(
+            replicaMetadataStore,
+            snapshotMetadataStore,
+            replicaCreationServiceConfig,
+            replicaEvictionServiceConfig,
+            meterRegistry);
+
+    int replicasCreated = replicaCreationService.createReplicasForUnassignedSnapshots();
+    int expectedReplicas = eligibleSnapshotsToCreate * replicasToCreate;
+    assertThat(replicasCreated).isEqualTo(expectedReplicas);
+    assertThat(MetricsUtil.getCount(ReplicaCreationService.REPLICAS_CREATED, meterRegistry))
+        .isEqualTo(expectedReplicas);
+    assertThat(MetricsUtil.getCount(ReplicaCreationService.REPLICAS_FAILED, meterRegistry))
+        .isZero();
+  }
+
+  @Test
+  public void shouldCreateReplicaWhenSnapshotAddedAfterRunning() throws Exception {
+    KaldbConfigs.ManagerConfig.ReplicaCreationServiceConfig replicaCreationServiceConfig =
+        KaldbConfigs.ManagerConfig.ReplicaCreationServiceConfig.newBuilder()
+            .setEventAggregationSecs(2)
+            .setReplicasPerSnapshot(2)
+            .setScheduleInitialDelayMins(0)
+            .setSchedulePeriodMins(10)
+            .build();
+
+    KaldbConfigs.ManagerConfig.ReplicaEvictionServiceConfig replicaEvictionServiceConfig =
+        KaldbConfigs.ManagerConfig.ReplicaEvictionServiceConfig.newBuilder()
+            .setReplicaLifespanMins(1440)
+            .build();
+
+    ReplicaCreationService replicaCreationService =
+        new ReplicaCreationService(
+            replicaMetadataStore,
+            snapshotMetadataStore,
+            replicaCreationServiceConfig,
+            replicaEvictionServiceConfig,
+            meterRegistry);
+    replicaCreationService.startAsync();
+    replicaCreationService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    assertThat(MetricsUtil.getCount(ReplicaCreationService.REPLICAS_CREATED, meterRegistry))
+        .isEqualTo(0);
+    assertThat(replicaMetadataStore.listSync().size()).isZero();
+    assertThat(replicaMetadataStore.getCached().size()).isZero();
+
+    // create a snapshot - we expect this to fire an event, and after the
+    // EventAggregationSecs duration, attempt to create the replicas
+    SnapshotMetadata snapshotA =
+        new SnapshotMetadata(
+            "a", "a", Instant.now().toEpochMilli() - 1, Instant.now().toEpochMilli(), 0, "a");
+    snapshotMetadataStore.createSync(snapshotA);
+
+    await().until(() -> replicaMetadataStore.getCached().size() == 2);
+    assertThat(MetricsUtil.getCount(ReplicaCreationService.REPLICAS_CREATED, meterRegistry))
+        .isEqualTo(2);
+    assertThat(MetricsUtil.getCount(ReplicaCreationService.REPLICAS_FAILED, meterRegistry))
+        .isZero();
+
+    replicaCreationService.stopAsync();
+    replicaCreationService.awaitTerminated(DEFAULT_START_STOP_DURATION);
+  }
+
+  @Test
+  public void shouldStillCreateReplicaIfFirstAttemptFails() throws Exception {
```

Comment:
Another way to test this would be to succeed in creating replicas once, then delete a replica and then run the assignment again to see if it we come back with a full list of replicas.